File: hackAstone_app/hackAstone_app/AuthStore.swift
import Foundation
import Combine

/// 与 Web `hackastone_auth` 对齐的登录会话。
struct AuthSession: Codable, Equatable {
    let token: String
    let userId: String
    let nickname: String
    var email: String?

    var displayName: String {
        if let email, !email.isEmpty { return nickname.isEmpty ? email : nickname }
        return nickname
    }
}

@MainActor
final class AuthStore: ObservableObject {
    static let storageKey = "hackastone_auth"

    @Published private(set) var session: AuthSession?

    var isLoggedIn: Bool { session != nil }

    static var bearerToken: String? {
        UserDefaults.standard.data(forKey: storageKey)
            .flatMap { try? JSONDecoder().decode(AuthSession.self, from: $0) }?
            .token
    }

    init() {
        load()
    }

    func load() {
        guard let data = UserDefaults.standard.data(forKey: Self.storageKey),
              let decoded = try? JSONDecoder().decode(AuthSession.self, from: data)
        else {
            session = nil
            return
        }
        session = decoded
    }

    func save(_ newSession: AuthSession) {
        session = newSession
        if let data = try? JSONEncoder().encode(newSession) {
            UserDefaults.standard.set(data, forKey: Self.storageKey)
        }
    }

    func clear() {
        session = nil
        UserDefaults.standard.removeObject(forKey: Self.storageKey)
    }

    func refreshProfile(userId: String, nickname: String, email: String?) {
        guard let current = session else { return }
        save(AuthSession(
            token: current.token,
            userId: userId,
            nickname: nickname,
            email: email
        ))
    }
}

File: hackAstone_Hams/entry/.preview/default/cache/default/default@PreviewArkTS/esmodule/debug/entry/src/main/ets/store/AuthStore.ts
import preferences from "@ohos:data.preferences";
/** 与 Web `hackastone_auth` 对齐的登录会话。 */
export interface AuthSession {
    token: string;
    userId: string;
    nickname: string;
    email?: string;
}
export class AuthSessionData implements AuthSession {
    token: string = '';
    userId: string = '';
    nickname: string = '';
    email?: string;
}
export function authSessionDisplayName(session: AuthSession): string {
    if (session.email !== undefined && session.email.length > 0) {
        return session.nickname.length === 0 ? session.email : session.nickname;
    }
    return session.nickname;
}
const PREF_FILE = 'hackastone_auth_store';
let appContext: Context | null = null;
export class AuthStore {
    static readonly storageKey: string = 'hackastone_auth';
    session: AuthSession | null = null;
    get isLoggedIn(): boolean {
        return this.session !== null;
    }
    static bindContext(ctx: Context): void {
        appContext = ctx;
    }
    private static prefs(): preferences.Preferences {
        if (appContext === null) {
            throw new Error('AuthStore: context not bound');
        }
        return preferences.getPreferencesSync(appContext, { name: PREF_FILE });
    }
    static get bearerToken(): string | null {
        try {
            const raw = AuthStore.prefs().getSync(AuthStore.storageKey, '') as string;
            if (raw.length === 0) {
                return null;
            }
            const decoded = JSON.parse(raw) as AuthSession;
            return decoded.token;
        }
        catch (_e) {
            return null;
        }
    }
    load(): void {
        try {
            const raw = AuthStore.prefs().getSync(AuthStore.storageKey, '') as string;
            if (raw.length === 0) {
                this.session = null;
                return;
            }
            this.session = JSON.parse(raw) as AuthSession;
        }
        catch (_e) {
            this.session = null;
        }
    }
    save(newSession: AuthSession): void {
        this.session = newSession;
        AuthStore.prefs().putSync(AuthStore.storageKey, JSON.stringify(newSession));
        AuthStore.prefs().flushSync();
    }
    clear(): void {
        this.session = null;
        AuthStore.prefs().deleteSync(AuthStore.storageKey);
        AuthStore.prefs().flushSync();
    }
    refreshProfile(userId: string, nickname: string, email: string | null): void {
        if (this.session === null) {
            return;
        }
        const current = this.session;
        const next = new AuthSessionData();
        next.token = current.token;
        next.userId = userId;
        next.nickname = nickname;
        if (email !== null && email.length > 0) {
            next.email = email;
        }
        this.save(next);
    }
}
